tuxemon/mod_manager.py

```diff
@@ -0,0 +1,190 @@
+import urllib.request
+
+from tuxemon.constants import paths
+from tuxemon.symlink_missing import symlink_missing
+import pathlib
+import os
+import json
+import zipfile
+import shutil
+
+class Manager:
+
+    def __init__(self, *other_urls, default_to_cache=True):
+        """
+        (basic) Mod managment library.
+        """
+        # TODO: Recreate the __init__
+        # TODO: Cache updates by default
+
+
+        if len(other_urls) == 0:
+            other_urls = ["http://127.0.0.1:5000"]
+
+        self.packages_path = os.path.join(paths.CACHE_DIR, "packages")
+
+        self.url = other_urls
+        self.packages = []
+
+        if default_to_cache:
+            self.packages = self.read_from_cache()
+
+    def write_to_cache(self):
+        """Writes self.packages to the cache file"""
+
+        with open(self.packages_path, "w") as file:
+            file.write(
+                json.dumps(self.packages, indent=4)
+            )
+    def read_from_cache(self):
+        """Read self.packages from the cache file"""
+        with open(self.packages_path) as file:
+            return json.loads(file.read())
+
+    def cache_to_pkglist(self):
+        """
+        Override self.packages with the cache.
+        Uses self.packages = self.read_from_cache()
+        """
+        self.packages = self.read_from_cache()
+
+    def update(self, url):
+        """Returns the response from the server"""
+        with urllib.request.urlopen(url + "/api/packages") as packages:
+            return json.loads(packages.read().decode("UTF-8"))
+
+    def update_all(self):
+        """
+        Updates all packages in self.package.
+        It automatically clears the self.package variable, and then
+        populates it from the data from the repositories.
+        """
+        self.packages = []
+
+        for i in self.url:
+            for package in self.update(i):
+                self.packages.append(package)
+
+        self.write_to_cache()
+
+    def list_packages(self):
+        """Returns package dictionary, either from the server or the cache"""
+        return self.packages
+
+    def download_package(self, name, release, repo=None, dont_extract=False, install_deps=True):
+        """Downloads the specified package"""
+        if repo is None:
+            repo = self.get_package_repo(name)
+
+            # Remove trailing slash
+            if repo[-1] == "/":
+                repo = repo[:-1]
+
+        url = str(repo) + f"/packages/{name}/releases/{str(release)}/download"
+        filename = os.path.join(paths.CACHE_DIR + f"/downloaded_packages/{name}.{release}.zip")
```

Comment:
this uses `+` inside of `os.path.join`.... i think you should be using a comma instead?

Reply:
also, there is no sanitation of the `name` value here.  a malformed package could be named something like `../foo` and it might be able to write files outside of the cache folder.